Add support for XDP UPS pods v2

ben-foxmoore opened the pull request:
@Wielewout a couple of things as I looked at merging this into the main UPS chart:

- I noticed you didn't have JAEGER_AGENT_HOST in the XDP Deployment spec. Do you know if this is no longer valid? I know the CU team switched to OpenTelemetry, but we still have this set in most of our K8S Pods/Deployments for the CU
- I've integrated the two charts, with one values file. My idea was you could specify both XDP interface settings as well as a number of non-XDP UPSs. 
  - Setting `numberOfUpStacks` to 0 would still create any XDP instances explicitly described in the values file, otherwise it will create any "missing" instances, i.e. `numberOfUpStacks - len(xdpUpsInterfaces)` extra instances (min 0 😁)
  -  Maybe it's just better to make it a simple `numberOfUpStacks = number of non-XDP instances` 🤷‍♂️ 
- I switched from a Deployment to a Pod. I'm not entirely sure why we did this previously, but at least it's consistent for now..
- Setting `XDP_OBJECT_FILE` in the Helm chart feels kind of odd, given that its already in a deterministic place at Docker build time. Could this already be set in the Dockerfile?
- Perhaps this is also the case for `HOSTMODE`? Its not quite clear what this does..
- Should we also consider adding node pinning from the list of XDP interfaces? If the Pod isn't pinned to a particular node, the interface details don't make any sense..

I haven't actually tested deploying this yet.. will do that next 😄

ben-foxmoore committed:
Add XDP-enabled UPS Pod for CU-UP

Wielewout commented:
> * I noticed you didn't have JAEGER_AGENT_HOST in the XDP Deployment spec. Do you know if this is no longer valid? I know the CU team switched to OpenTelemetry, but we still have this set in most of our K8S Pods/Deployments for the CU

JAEGER_AGENT_HOST has been out for a while now, so it doesn't do anything.

> * I've integrated the two charts, with one values file. My idea was you could specify both XDP interface settings as well as a number of non-XDP UPSs.
>   
>   * Setting `numberOfUpStacks` to 0 would still create any XDP instances explicitly described in the values file, otherwise it will create any "missing" instances, i.e. `numberOfUpStacks - len(xdpUpsInterfaces)` extra instances (min 0 😁)
>   * Maybe it's just better to make it a simple `numberOfUpStacks = number of non-XDP instances` 🤷‍♂️

I think there are a few issues with using a single chart (or at least with not having the values being repeated with differentiation for xdp and regular). As mentioned later you would need to pin the XDP ups to a specific node with the configured interface, so a separate node selector would need to be introduced to differentiate this from the regular UPS. In the same fashion I think we should decouple all configuration options (like resources etc.) so they can be specified. Hence another subchart seems more logical to me.

> * I switched from a Deployment to a Pod. I'm not entirely sure why we did this previously, but at least it's consistent for now..

I'm now thinking maybe a DaemonSet is more appropriate as we're working on nodes (sure we would select only a single one, but as scaling is not really the purpose it seems more appropriate; also xdp ups containers will fail if they are the second on the interface). Not in the scope for now though.

> * Setting `XDP_OBJECT_FILE` in the Helm chart feels kind of odd, given that its already in a deterministic place at Docker build time. Could this already be set in the Dockerfile?

I've added that to the Dockerfile for the new charts as I felt the same.

> * Perhaps this is also the case for `HOSTMODE`? Its not quite clear what this does..

It's for the XDP module to know if it needs to do for example a rewrite of the source address. In this case I think it's appropriately placed (as a necessary option to enable within kubernetes).

> * Should we also consider adding node pinning from the list of XDP interfaces? If the Pod isn't pinned to a particular node, the interface details don't make any sense..

Indeed a very important one. I suppose we should see the node and (interface or ip address) as some sort of single key. Better would be to determine the interface based on the address, something I can still add maybe to the docker images and then with a patch release.

> I haven't actually tested deploying this yet.. will do that next 😄

ben-foxmoore commented:
> > * I noticed you didn't have JAEGER_AGENT_HOST in the XDP Deployment spec. Do you know if this is no longer valid? I know the CU team switched to OpenTelemetry, but we still have this set in most of our K8S Pods/Deployments for the CU
> 
> JAEGER_AGENT_HOST has been out for a while now, so it doesn't do anything.

Do you know if there's a replacement that we should be setting?

> > * I've integrated the two charts, with one values file. My idea was you could specify both XDP interface settings as well as a number of non-XDP UPSs.
> >   
> >   * Setting `numberOfUpStacks` to 0 would still create any XDP instances explicitly described in the values file, otherwise it will create any "missing" instances, i.e. `numberOfUpStacks - len(xdpUpsInterfaces)` extra instances (min 0 😁)
> >   * Maybe it's just better to make it a simple `numberOfUpStacks = number of non-XDP instances` 🤷‍♂️
> 
> I think there are a few issues with using a single chart (or at least with not having the values being repeated with differentiation for xdp and regular). [..] In the same fashion I think we should decouple all configuration options (like resources etc.) so they can be specified. Hence another subchart seems more logical to me.

Aren't we going to be moving away from lots of individual charts in the short term future? Adding a new one at this point seems a little pointless in that case. Also, I imagine we won't actually be deploying both XDP and non-XDP UPS instances, so in practice the configurations (such as resources) can be set as needed depending on the scenario?

Thinking about it again , I'll change to the simple `numberOfUpStacks = number of non-XDP instances`. Most of the time, I expect this will be 0 once XDP is supported properly, and it's probably more clear that if you set it to any other number, that's how many non-XDP instances you'll get.




> > * Setting `XDP_OBJECT_FILE` in the Helm chart feels kind of odd, given that its already in a deterministic place at Docker build time. Could this already be set in the Dockerfile?
> 
> I've added that to the Dockerfile for the new charts as I felt the same.

Is this already in the CU 4.2 images? Can I leave it out then?

> > * Perhaps this is also the case for `HOSTMODE`? Its not quite clear what this does..
> 
> It's for the XDP module to know if it needs to do for example a rewrite of the source address. In this case I think it's appropriately placed (as a necessary option to enable within kubernetes).

OK 👍 

> 
> > * Should we also consider adding node pinning from the list of XDP interfaces? If the Pod isn't pinned to a particular node, the interface details don't make any sense..
> 
> Indeed a very important one. I suppose we should see the node and (interface or ip address) as some sort of single key. Better would be to determine the interface based on the address, something I can still add maybe to the docker images and then with a patch release.

> > * I switched from a Deployment to a Pod. I'm not entirely sure why we did this previously, but at least it's consistent for now..
> 
> I'm now thinking maybe a DaemonSet is more appropriate as we're working on nodes (sure we would select only a single one, but as scaling is not really the purpose it seems more appropriate; also xdp ups containers will fail if they are the second on the interface). Not in the scope for now though.

> As mentioned later you would need to pin the XDP ups to a specific node with the configured interface, so a separate node selector would need to be introduced to differentiate this from the regular UPS.

I'm not sure DaemonSet is appropriate - that's for something you specifically want running on each Node of the Cluster, but in this case, we want specific Pods running on specific Nodes.

The pinning needs to be correlated to the interface name/IP, so I think I need to add this to the `.Values.xdp.interfaces` list:

```yaml
interfaces:
- name: "eth0"
  address: "127.0.0.1"
  nodeSelector:
  - example-label: example-value
  nodeName: example-node-name # apparently this is also possible
```

Alternatively something more generic like:
```yaml
interfaces:
- name: "eth0"
  address: "127.0.0.1"
  extraPodYaml:
      nodeSelector:
      - example-label: example-value
      nodeName: example-node-name # apparently this is also possible
```

We should probably also account for the `draxNodeSelector` because if this is being used, then you could reasonably expect that the XDP UPS is pinned to that node if no other selectors are specified...

Wielewout commented:
> > > * I noticed you didn't have JAEGER_AGENT_HOST in the XDP Deployment spec. Do you know if this is no longer valid? I know the CU team switched to OpenTelemetry, but we still have this set in most of our K8S Pods/Deployments for the CU
> > 
> > 
> > JAEGER_AGENT_HOST has been out for a while now, so it doesn't do anything.
> 
> Do you know if there's a replacement that we should be setting?

No, there isn't.

> > > * I've integrated the two charts, with one values file. My idea was you could specify both XDP interface settings as well as a number of non-XDP UPSs.
> > >   
> > >   * Setting `numberOfUpStacks` to 0 would still create any XDP instances explicitly described in the values file, otherwise it will create any "missing" instances, i.e. `numberOfUpStacks - len(xdpUpsInterfaces)` extra instances (min 0 😁)
> > >   * Maybe it's just better to make it a simple `numberOfUpStacks = number of non-XDP instances` 🤷‍♂️
> > 
> > 
> > I think there are a few issues with using a single chart (or at least with not having the values being repeated with differentiation for xdp and regular). [..] In the same fashion I think we should decouple all configuration options (like resources etc.) so they can be specified. Hence another subchart seems more logical to me.
> 
> Aren't we going to be moving away from lots of individual charts in the short term future? Adding a new one at this point seems a little pointless in that case. Also, I imagine we won't actually be deploying both XDP and non-XDP UPS instances, so in practice the configurations (such as resources) can be set as needed depending on the scenario?
>
> Thinking about it again , I'll change to the simple `numberOfUpStacks = number of non-XDP instances`. Most of the time, I expect this will be 0 once XDP is supported properly, and it's probably more clear that if you set it to any other number, that's how many non-XDP instances you'll get.

It's not because we won't have multiple subcharts that we can't have a clear separation between deployed microservices. From the top chart I think it also looks clearer if you see two separate maps for them.

As for not allowing the two mixed together, I would then enforce it by explicitly adding that it's required for the the list length to be `0` or `numOfUpStacks`. In that case I would be fine with it. It also allows for less bad configurations maybe. (That would also mean not doing the simpler thing then. Otherwise the check should be that one of them is 0.)
 
> > > * Setting `XDP_OBJECT_FILE` in the Helm chart feels kind of odd, given that its already in a deterministic place at Docker build time. Could this already be set in the Dockerfile?
> > 
> > 
> > I've added that to the Dockerfile for the new charts as I felt the same.
> 
> Is this already in the CU 4.2 images? Can I leave it out then?

Yes

> > > * Should we also consider adding node pinning from the list of XDP interfaces? If the Pod isn't pinned to a particular node, the interface details don't make any sense..
> > 
> > 
> > Indeed a very important one. I suppose we should see the node and (interface or ip address) as some sort of single key. Better would be to determine the interface based on the address, something I can still add maybe to the docker images and then with a patch release.
> 
> > > * I switched from a Deployment to a Pod. I'm not entirely sure why we did this previously, but at least it's consistent for now..
> > 
> > 
> > I'm now thinking maybe a DaemonSet is more appropriate as we're working on nodes (sure we would select only a single one, but as scaling is not really the purpose it seems more appropriate; also xdp ups containers will fail if they are the second on the interface). Not in the scope for now though.
> 
> > As mentioned later you would need to pin the XDP ups to a specific node with the configured interface, so a separate node selector would need to be introduced to differentiate this from the regular UPS.
> 
> I'm not sure DaemonSet is appropriate - that's for something you specifically want running on each Node of the Cluster, but in this case, we want specific Pods running on specific Nodes.
> 
> The pinning needs to be correlated to the interface name/IP, so I think I need to add this to the `.Values.xdp.interfaces` list:
> 
> ```yaml
> interfaces:
> - name: "eth0"
>   address: "127.0.0.1"
>   nodeSelector:
>   - example-label: example-value
>   nodeName: example-node-name # apparently this is also possible
> ```
> 
> Alternatively something more generic like:
> 
> ```yaml
> interfaces:
> - name: "eth0"
>   address: "127.0.0.1"
>   extraPodYaml:
>       nodeSelector:
>       - example-label: example-value
>       nodeName: example-node-name # apparently this is also possible
> ```
> 
> We should probably also account for the `draxNodeSelector` because if this is being used, then you could reasonably expect that the XDP UPS is pinned to that node if no other selectors are specified...

I really don't like the `draxNodeSelector` as it makes things unnecessary complex, while it is just another node selector...

Anyway, I like the first one more (with the nodeName only at least), but in general I would want to drop the interface name as that can be looked up (we're in the host network anyway for XDP to work). We can maybe even put this in an init container if we really don't want to add a patch release to the CU (though I would prefer the patch release):

```sh
ip -br a | grep "<ip>" | cut -d " " -f1
```

Then we could leave the interfaces like this (we could still allow the interface to be set, but optional, like mtu):

```yaml
interfaces:
- nodeName: "example-node-name"
  address: "127.0.0.1"
```

ben-foxmoore committed:
Don't allow XDP and non-XDP UPS at the same time

ben-foxmoore committed:
Allow pinning XDP UPS instances to specific nodes

ben-foxmoore committed:
Use entrypoint.sh in CU-UP UPS Pods

ben-foxmoore commented:
> > > JAEGER_AGENT_HOST has been out for a while now, so it doesn't do anything.
> > 
> > 
> > Do you know if there's a replacement that we should be setting?
> 
> No, there isn't.

OK 👍
 
> As for not allowing the two mixed together, I would then enforce it by explicitly adding that it's required for the the list length to be `0` or `numOfUpStacks`. In that case I would be fine with it. It also allows for less bad configurations maybe. (That would also mean not doing the simpler thing then. Otherwise the check should be that one of them is 0.)

I think that's a good idea - I'll try to write a helper template that validates that.

> > > > * Setting `XDP_OBJECT_FILE` in the Helm chart feels kind of odd, given that its already in a deterministic place at Docker build time. Could this already be set in the Dockerfile?
> > > 
> > > 
> > > I've added that to the Dockerfile for the new charts as I felt the same.
> > 
> > 
> > Is this already in the CU 4.2 images? Can I leave it out then?
> 
> Yes

OK - will remove this from the chart then.

> > We should probably also account for the `draxNodeSelector` because if this is being used, then you could reasonably expect that the XDP UPS is pinned to that node if no other selectors are specified...
> 
> I really don't like the `draxNodeSelector` as it makes things unnecessary complex, while it is just another node selector...

It's there so that in the RIC values you can specify the NodeSelector once in a global value, and not have to specify it multiple times, once for each chart. (i.e. the RIC default values all reference the global value via templating `"{{ .Values.global.draxNodeSelector }}"`). I don't love it either but I don't see an easy way to replace it, at least while we still have many subcharts.

> > The pinning needs to be correlated to the interface name/IP, so I think I need to add this to the `.Values.xdp.interfaces` list:
> > ```yaml
> > interfaces:
> > - name: "eth0"
> >   address: "127.0.0.1"
> >   nodeSelector:
> >   - example-label: example-value
> >   nodeName: example-node-name # apparently this is also possible
> > ```
> >   
> > Alternatively something more generic like:
> > ```yaml
> > interfaces:
> > - name: "eth0"
> >   address: "127.0.0.1"
> >   extraPodYaml:
> >       nodeSelector:
> >       - example-label: example-value
> >       nodeName: example-node-name # apparently this is also possible
> > ```
> Anyway, I like the first one more (with the nodeName only at least), but in general I would want to drop the interface name as that can be looked up (we're in the host network anyway for XDP to work). We can maybe even put this in an init container if we really don't want to add a patch release to the CU (though I would prefer the patch release):
> 
> ```shell
> ip -br a | grep "<ip>" | cut -d " " -f1
> ```
> 
> Then we could leave the interfaces like this (we could still allow the interface to be set, but optional, like mtu):
> 
> ```yaml
> interfaces:
> - nodeName: "example-node-name"
>   address: "127.0.0.1"
> ```

I can imagine that at some point it might be nice/feasible to assign the IP to the specified interface during deployment/installation time, avoiding having to manually pre-configure it. In that case, we'd need both IP and interface, so maybe better to have both already?

Wielewout commented:
> > > We should probably also account for the `draxNodeSelector` because if this is being used, then you could reasonably expect that the XDP UPS is pinned to that node if no other selectors are specified...
> > 
> > 
> > I really don't like the `draxNodeSelector` as it makes things unnecessary complex, while it is just another node selector...
> 
> It's there so that in the RIC values you can specify the NodeSelector once in a global value, and not have to specify it multiple times, once for each chart. (i.e. the RIC default values all reference the global value via templating `"{{ .Values.global.draxNodeSelector }}"`). I don't love it either but I don't see an easy way to replace it, at least while we still have many subcharts.

We could simplify it at least, but a discussion for the new chart releases perhaps. At the moment we'll have to keep it anyways for consistency between all the charts...

> > > The pinning needs to be correlated to the interface name/IP, so I think I need to add this to the `.Values.xdp.interfaces` list:
> > > ```yaml
> > > interfaces:
> > > - name: "eth0"
> > >   address: "127.0.0.1"
> > >   nodeSelector:
> > >   - example-label: example-value
> > >   nodeName: example-node-name # apparently this is also possible
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Alternatively something more generic like:
> > > ```yaml
> > > interfaces:
> > > - name: "eth0"
> > >   address: "127.0.0.1"
> > >   extraPodYaml:
> > >       nodeSelector:
> > >       - example-label: example-value
> > >       nodeName: example-node-name # apparently this is also possible
> > > ```
> > 
> > 
> > Anyway, I like the first one more (with the nodeName only at least), but in general I would want to drop the interface name as that can be looked up (we're in the host network anyway for XDP to work). We can maybe even put this in an init container if we really don't want to add a patch release to the CU (though I would prefer the patch release):
> > ```shell
> > ip -br a | grep "<ip>" | cut -d " " -f1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then we could leave the interfaces like this (we could still allow the interface to be set, but optional, like mtu):
> > ```yaml
> > interfaces:
> > - nodeName: "example-node-name"
> >   address: "127.0.0.1"
> > ```
> 
> I can imagine that at some point it might be nice/feasible to assign the IP to the specified interface during deployment/installation time, avoiding having to manually pre-configure it. In that case, we'd need both IP and interface, so maybe better to have both already?

Not sure if the assignment will really be possible, but a future improvement anyways as it's not part of the images in this release.

ben-foxmoore commented:
@Wielewout I ended up with this for only allowing either XDP or non-XDP:
```helm
{{- if and (gt (len .Values.xdp.interfaces) 0) (gt (tpl .Values.numberOfUpStacks . | int) 0) -}}
{{- fail "UPS XDP Interfaces can only be specified if numberOfUpStacks is set to 0" -}}
{{- end -}}
```

I added support for `nodeName`/`nodeSelector`, as well as adding `nodeSelectorOverride`. Given that placement of this Pod is even more important that normal, I thought that might be useful and I couldn't see a way to achieve overriding the global `nodeSelector`/`draxNodeSelector` values otherwise.

I think this is ready to go now, unless you see any other issues with it.

Wielewout commented:
Looks good :+1:
